### ISSUE TITLE
Permanently deleted media files

### ISSUE THREAD

**Issue by larssn** [humanmade/S3-Uploads#92]
Hey, good plugin, works well.

Quick question that may or may not be an issue:
When deleting a media file, the originally uploaded file gets left behind - only the different thumbnail variants gets deleted. Is this intended behavior?

We run a multisite setup, and would prefer that all files permanently deleted through Wordpress would get removed completely from S3.

Thanks.

Ps. I can delete the left over file through the CLI, so at least that is working, although not convenient:
wp s3-uploads rm uploads/2016/04/image.jpg --debug --url=www.our-multisite-setup.com

**Comment by larssn** [comment]
Further strangeness: I remove the wp_delete_file filter from S3_Uploads: Then the original file gets deleted, but the thumbnail variants doesn't. I can understand why they dont get deleted, but now I don't understand why the original file DOES get deleted... 🎱

**Comment by joehoyle** [comment]
Oops think I broke this in https://github.com/humanmade/S3-Uploads/pull/87 whereby I fixed the removing of sizes, but that apparently broke the original

**Comment by larssn** [comment]
Where does the delete hook occur? Been looking through the code, and all I see is the rm function in the CLI class (which seems irrelevant), and the wp delete filter that you added.

**Comment by joehoyle** [comment]
WordPress deletes the file internally, and just passes it through the `wp_delete_file` hook first, which I tinkered with. Will get a patch up for this.

**Comment by larssn** [comment]
👍
Thanks for the help!

**Comment by larssn** [comment]
Forgive me, I'm new when it comes to navigating github. Is there a section for patches? Been looking around for it.

**Comment by thefrosty** [comment]
Will this get patched in `1.1.1` or in the `2.x` release?

**Comment by nuvoPoint** [comment]
We are having the same issue. Is there a fix for this or should we make a fork?

**Comment by josephabrahams** [comment]
:+1:

**Comment by mathewemoore** [comment]
Is anyone working on a fix for this? I am working on a project and everything is working fine except for deleting the original file.

**Comment by MordiSacks** [comment]
I went through the whole project, for the life of me I can't understand why main file does not get deleted.

**Comment by ikappas** [comment]
Hi all,

When an attachment is deleted the function `wp_delete_attachment` is called that checks whether the attachment has thumbnail, intermmediate and backup images and removes them with the following code:

```php
$thumbfile = apply_filters( 'wp_delete_file', $thumbfile );
@ unlink( path_join($uploadpath['basedir'], $thumbfile) );
```
From the snippet you can see that the `wp_delete_file` filter is applied and then the path is joined back together with the $uploadpath['basedir'] yielding in effect the original absolute s3 path. 

When these iterations are completed `wp_delete_attachment` finally calls `wp_delete_file` that attempts to remove the main attachment file with the following code:
```php
	$delete = apply_filters( 'wp_delete_file', $file );
	if ( ! empty( $delete ) ) {
		@unlink( $delete );
	}
```
As you can see this one requires an **absolute** file path and not a **relative** file, thus failing to delete the main attachment file.

The only way to fix this issue is to implement a check in class-s3-uploads::wp_filter_delete_file that checks the call-stack to determine whether the call originates from  `wp_delete_attachment` or `wp_delete_file` and return the appropriate filename.

Hope this helps!

**Comment by ikappas** [comment]
And here is a solution:
```php
	/**
	 * When WordPress removes files, it's expecting to do so on
	 * absolute file paths, as such it breaks when using uris for
	 * file paths (such as s3://...). We have to filter the file_path
	 * to only return the relative section, to play nice with WordPress
	 * handling.
	 *
	 * @param  string $file_path
	 * @return string
	 */
	public function wp_filter_delete_file( $file_path ) {
		$dir = wp_upload_dir();

		if ( $this->func_caller_is( 'wp_delete_file' ) ) {
			return $file_path;
		}

		return str_replace( trailingslashit( $dir['basedir'] ), '', $file_path );
	}

	/**
	 * Determine whether the specified function exists in the call stack.
	 *
	 * @param string $function The function name to check.
	 * @return bool True if the specified function is in the call stack; Otherwise false.
	 */
	protected function func_caller_is( $function ) {
		$stack = debug_backtrace();
		$functions = wp_list_pluck( $stack, 'function' );
		return in_array( $function, array_unique( $functions ) );
	}
```

**Comment by thefrosty** [comment]
@ikappas You should open a PR.

**Comment by rmccue** [comment]
@ikappas This seems like a bug that should be filed with WordPress itself; can you create a ticket [on Trac](https://core.trac.wordpress.org/) please?

I'd rather we avoid using a backtrace if at all possible. Is there any other way to detect this?

**Comment by ikappas** [comment]
@rmccue It sure looks like a bug but I wont get into opening a Track ticket. Feel free to do so yourself but take a look at https://core.trac.wordpress.org/ticket/17864 for a similar issue and look at how long it took to fix.

Please don't get me wrong, a Trac ticket should be opened for this. Meanwhile, leaving deleted artifacts behind on S3 is not a good practice.

This issue has been pending here since last April. If you choose to wait for a Track ticket to **maybe** be resolved via WordPress Core... do the math.

Using backtrace is hacky but gets the job done and I don't know of any other way to go about this.

**Comment by mathewemoore** [comment]
@ikappas I tested your modification and it worked smashingly on all of my websites using this plugin. Hopefully, this gets fixed in wordpress core, but if not, it's not a huge deal to me. I'm just happy everything gets cleaned up.
[http://www.wismusic.com/](http://www.wismusic.com/)

**Comment by ikappas** [comment]
@rmccue Do we have an updated estimate on this issue now that 4.8 is out ?

**Comment by jeremyfelt** [comment]
I ***think*** #215 is a valid enough way forward here without using backtrace until fixed upstream in core.

**Comment by joehoyle** [comment]
This should have been fixed by #215
